design-proposals/instancetype.kubevirt.io/custom-resource-sizing.md

```diff
@@ -0,0 +1,163 @@
+
+## Non Goals
+
+## User Stories
+
+* As a VM owner I want to provide my own vCPU and memory resource sizing while using commonly deployed instance types  
+* As a VM owner I want to create customized versions of commonly deployed instance types and preferences specific to my workload use case  
+* As a cluster admin I want users to be able to provide their own vCPU and memory resource sizing when using commonly deployed instance types  
+* As a cluster admin I want to create customized versions of commonly deployed instance types and preferences specific to my workload use case
+
+## Repos
+
+* `kubevirt/kubevirt`  
+* `kubevirt/common-instancetypes`
+
+# Design
+
+## Support for user supplied vCPU and memory instance type resource sizing
+
+The original design of the instance type API and CRDs required that vCPU and memory resource sizing was always provided when creating an instance type.
+
+This obviously constrained users to the resource sizing already baked into commonly deployed instance types within their environment.
+
+It would be more flexible if the resource sizing within an instance type was optional and could instead be provided by the users VirtualMachine.
+
+### Make `spec.cpu.guest` and `spec.memory.guest` optional
+
+The first step here is to make both the amount of vCPUs and memory provided by an instance type optional. This behavioral change will require a new version of the API, `instancetype.kubevirt.io/v1beta2`.
+
+### Introduce `spec.template.spec.domain.cpu.guest`
+
+This aims to align the VirtualMachine and instance type APIs in terms of expressing the number of vCPUs of provide to the eventual guest.
+
+Until now the VM could only express an amount of vCPU resource through resource requests or through the guest visible CPU topology with the number of vCPUs being the total of sockets\*cores\*threads.
+
+By introducing a standalone `guest` attribute that expresses a number of vCPUs we can also allow this value to be used by any associated preference when determining the guest visible CPU topology.
+
+Without the use of a preference the vCPUs provided by this field will be made visible to the guest OS as single core sockets by default.
+
+This field would be the only supported way of providing a number of vCPUs using the VM when also using an instance type without a set amount of vCPUs.
+
+The field would also be mutually exclusive with the original `spec.template.spec.domain.cpu.{cores,sockets,threads}` fields.
+
+### Provide a `custom` size for each common-instancetypes instance type class
+
+A new `custom` size would be introduced for all but the universal instance type class without vCPUs or memory defined allowing users to provide their own values through the VirtualMachine.
+
+A `custom` size obviously not making sense for the universal instance type class that only provides vCPUs and memory.
+
+## Allow for the easy creation of customization common-instancetype resources
+
+### Introduce `virtctl customize`
```

Comment:
Why? Can't users just `kubectl edit` to make changes?

Reply:
The use case with `virtctl customize` is that we want to start with an object in the cluster, make some small changes, automate kind conversion/metadata removal and then output the newly crafted definition.

Does `kubectl edit` actually support taking an existing object and only outputting the result locally? I thought it always attempted to mutate the object in the cluster?